File: backend/src/services/plugins/github.ts
/**
 * Plugin Platform — GitHub fetcher
 *
 * Downloads a plugin from a GitHub repository using the public API:
 *   1. Resolve the default branch
 *   2. Fetch the recursive file tree
 *   3. Download each file's content (raw.githubusercontent.com)
 * This needs no auth token and works for public repos.
 */

import { logger as appLogger } from '../logger';

const GITHUB_API = 'https://api.github.com';
const RAW_BASE = 'https://raw.githubusercontent.com';
const USER_AGENT = 'Kasalix-Server/1.0';

export interface RepoSpec {
  owner: string;
  repo: string;
  /** Optional subdirectory inside the repo that holds the plugin */
  path?: string;
  branch: string;
}

/** Normalize "owner/repo", "owner/repo/subdir", or a full GitHub URL. */
export function parseRepoInput(input: string): { owner: string; repo: string; path?: string } {
  let s = input.trim();
  if (!s) throw new Error('Repository is required');
  // Full URL form: https://github.com/owner/repo/tree/branch/subdir or /owner/repo
  const urlMatch = s.match(/^https?:\/\/github\.com\/([^/\s]+)\/([^/\s#?]+)(?:\/(?:tree|blob)\/[^/\s]+)?(?:\/([^#?\s]*))?/i);
  let parts: string[];
  if (urlMatch) {
    parts = [urlMatch[1], urlMatch[2], ...(urlMatch[3] ? urlMatch[3].split('/').filter(Boolean) : [])];
  } else {
    parts = s.split('/').filter(Boolean);
  }
  if (parts.length < 2) {
    throw new Error('Invalid repository. Use "owner/repo" or a GitHub URL.');
  }
  const owner = parts[0].replace(/[^\w.-]/g, '');
  const repo = parts[1].replace(/[^\w.-]/g, '');
  if (!owner || !repo) throw new Error('Invalid repository name');
  const path = parts.length > 2 ? parts.slice(2).join('/') : undefined;
  return { owner, repo, path: path || undefined };
}

async function githubJson(url: string): Promise<any> {
  const res = await fetch(url, {
    headers: {
      'User-Agent': USER_AGENT,
      Accept: 'application/vnd.github+json',
      'X-GitHub-Api-Version': '2022-11-28',
    },
  });
  if (res.status === 403 || res.status === 429) {
    throw new Error('GitHub API rate limit reached. Try again in a few minutes.');
  }
  if (!res.ok) {
    throw new Error(`GitHub API error ${res.status} for ${url}`);
  }
  return res.json();
}

/** Resolve the repo's default branch and validate it exists. */
export async function resolveRepo(spec: { owner: string; repo: string }): Promise<string> {
  const info = await githubJson(`${GITHUB_API}/repos/${spec.owner}/${spec.repo}`);
  const branch = info?.default_branch || 'main';
  if (typeof branch !== 'string' || !branch) {
    throw new Error('Could not determine the repository default branch');
  }
  return branch;
}

/** Get every file path in the repo (optionally under a subdirectory). */
export async function getRepoFiles(
  owner: string,
  repo: string,
  branch: string,
  subdir?: string
): Promise<string[]> {
  const tree = await githubJson(
    `${GITHUB_API}/repos/${owner}/${repo}/git/trees/${encodeURIComponent(branch)}?recursive=1`
  );
  const entries: { path?: string; type?: string }[] = tree?.tree || [];
  const prefix = subdir ? subdir.replace(/^\/+|\/+$/g, '') + '/' : '';
  return entries
    .filter((e) => e.type === 'blob' && e.path)
    .map((e) => e.path as string)
    .filter((p) => {
      if (prefix && !p.startsWith(prefix)) return false;
      const rel = prefix ? p.slice(prefix.length) : p;
      // Skip VCS metadata and build junk
      if (rel.startsWith('.git/') || rel.startsWith('node_modules/') || rel === '.git') return false;
      if (rel.includes('/node_modules/')) return false;
      return true;
    });
}

/** Download a single file's text content. Returns null if not found. */
export async function downloadRawFile(
  owner: string,
  repo: string,
  branch: string,
  pathInRepo: string
): Promise<string | null> {
  const url = `${RAW_BASE}/${owner}/${repo}/${encodeURIComponent(branch)}/${pathInRepo
    .split('/')
    .map((seg) => encodeURIComponent(seg))
    .join('/')}`;
  const res = await fetch(url, { headers: { 'User-Agent': USER_AGENT } });
  if (res.status === 404) return null;
  if (!res.ok) throw new Error(`Failed to download ${pathInRepo} (HTTP ${res.status})`);
  const text = await res.text();
  if (text.length > 5 * 1024 * 1024) {
    throw new Error(`File too large: ${pathInRepo} (over 5 MB)`);
  }
  return text;
}

/** Download the repo metadata we need in one call (used by the install flow). */
export async function fetchRepoTree(spec: RepoSpec): Promise<string[]> {
  return getRepoFiles(spec.owner, spec.repo, spec.branch, spec.path);
}

export { appLogger };

File: backend/app/services/plugins/github.py
"""Plugin Platform — GitHub fetcher (mirrors backend/src/services/plugins/github.ts).

Downloads a plugin from a GitHub repository using the public API:
  1. Resolve the default branch
  2. Fetch the recursive file tree
  3. Download each file's content (raw.githubusercontent.com)
This needs no auth token and works for public repos.
"""

from __future__ import annotations

import re
import urllib.parse
from typing import Any

import httpx

from ...logger import error as log_error, warn as log_warn

GITHUB_API = "https://api.github.com"
RAW_BASE = "https://raw.githubusercontent.com"
USER_AGENT = "Kasalix-Server/1.0"

_GITHUB_CLIENT: httpx.AsyncClient | None = None


def _client() -> httpx.AsyncClient:
    global _GITHUB_CLIENT
    if _GITHUB_CLIENT is None:
        _GITHUB_CLIENT = httpx.AsyncClient(
            timeout=20,
            headers={
                "User-Agent": USER_AGENT,
                "Accept": "application/vnd.github+json",
                "X-GitHub-Api-Version": "2022-11-28",
            },
        )
    return _GITHUB_CLIENT


def parse_repo_input(input: str) -> dict[str, str]:
    """Normalize 'owner/repo', 'owner/repo/subdir', or a full GitHub URL."""
    s = input.strip()
    if not s:
        raise ValueError("Repository is required")
    url_match = re.match(
        r"^https?://github\.com/([^\s/]+)/([^\s#?]+)(?:/(?:tree|blob)/[^\s/]+)?(?:/([^#?\s]*))?",
        s,
        re.I,
    )
    if url_match:
        parts = [url_match.group(1), url_match.group(2)]
        if url_match.group(3):
            parts.extend(p for p in url_match.group(3).split("/") if p)
    else:
        parts = [p for p in s.split("/") if p]
    if len(parts) < 2:
        raise ValueError('Invalid repository. Use "owner/repo" or a GitHub URL.')
    owner = re.sub(r"[^\w.-]", "", parts[0])
    repo = re.sub(r"[^\w.-]", "", parts[1])
    if not owner or not repo:
        raise ValueError("Invalid repository name")
    path = "/".join(parts[2:]) if len(parts) > 2 else ""
    return {"owner": owner, "repo": repo, "path": path or ""}


async def _github_json(url: str) -> Any:
    client = _client()
    res = await client.get(url)
    if res.status_code in (403, 429):
        raise RuntimeError("GitHub API rate limit reached. Try again in a few minutes.")
    if res.status_code >= 400:
        raise RuntimeError(f"GitHub API error {res.status_code} for {url}")
    return res.json()


async def resolve_repo(spec: dict[str, str]) -> str:
    """Resolve the repo's default branch and validate it exists."""
    info = await _github_json(f"{GITHUB_API}/repos/{spec['owner']}/{spec['repo']}")
    branch = info.get("default_branch") or "main"
    if not isinstance(branch, str) or not branch:
        raise RuntimeError("Could not determine the repository default branch")
    return branch


async def get_repo_files(owner: str, repo: str, branch: str, subdir: str | None = None) -> list[str]:
    """Get every file path in the repo (optionally under a subdirectory)."""
    tree = await _github_json(
        f"{GITHUB_API}/repos/{owner}/{repo}/git/trees/{urllib.parse.quote(branch)}?recursive=1"
    )
    entries = tree.get("tree") or []
    prefix = (subdir or "").strip("/") + "/" if subdir else ""
    result: list[str] = []
    for e in entries:
        if e.get("type") != "blob" or not e.get("path"):
            continue
        p: str = e["path"]
        if prefix and not p.startswith(prefix):
            continue
        rel = p[len(prefix):] if prefix else p
        if rel == ".git" or rel.startswith(".git/") or rel.startswith("node_modules/") or "/node_modules/" in rel:
            continue
        result.append(p)
    return result


async def download_raw_file(owner: str, repo: str, branch: str, path_in_repo: str) -> str | None:
    """Download a single file's text content. Returns None if not found."""
    encoded_path = "/".join(urllib.parse.quote(seg) for seg in path_in_repo.split("/"))
    url = f"{RAW_BASE}/{owner}/{repo}/{urllib.parse.quote(branch)}/{encoded_path}"
    client = httpx.AsyncClient(timeout=20, headers={"User-Agent": USER_AGENT})
    try:
        res = await client.get(url)
    finally:
        await client.aclose()
    if res.status_code == 404:
        return None
    if res.status_code >= 400:
        raise RuntimeError(f"Failed to download {path_in_repo} (HTTP {res.status_code})")
    text = res.text
    if len(text) > 5 * 1024 * 1024:
        raise RuntimeError(f"File too large: {path_in_repo} (over 5 MB)")
    return text


async def fetch_repo_tree(spec: dict[str, Any]) -> list[str]:
    return await get_repo_files(spec["owner"], spec["repo"], spec["branch"], spec.get("path") or None)